### ISSUE TITLE
SSL passthrough for Ingress container

### ISSUE THREAD

**Issue by sandys** [kubernetes/contrib#1854]
the current ingress container cannot do L4 routing. I dont want to terminate my ssl at the ingress container because my ssl config is complicated (with lua rules and all) and i would rather not port them over to annotations.

I see that there is a readme for [proxy protocol](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx/examples/proxy-protocol) as well plans to use [Caddy](https://github.com/kubernetes/kubernetes/issues/19899) as the ingress container.

 Is there a long term plan to add SSL pass through for the ingress container ?

**Comment by aledbf** [comment]
> Is there a long term plan to add SSL pass through for the ingress container ?

@sandys yes, please check https://github.com/kubernetes/kubernetes/issues/24145
That being said features like "SSL pass through" must be supported by the software that provides the reverse proxy (for instance this is not possible with nginx)

**Comment by Sushant** [comment]
> That being said features like "SSL pass through" must be supported by the software that provides the reverse proxy (for instance this is not possible with nginx)

@aledbf NGINX Ingress controller (nginx-ingress-controller:0.8.3) seems to be running `nginx/1.11.3`. I believe this would be possible with `nginx/1.11.5` which has the `ngx_stream_ssl_preread_module`?

**Comment by aledbf** [comment]
@Sushant yes and no. It is possible **IF** you use the new stream feature but if you do that means TCP only because can't do L7 balancing without terminating TLS (ie all the Ingress with TLS termination in the controller won't work).

Check this issue https://github.com/nginxinc/kubernetes-ingress/issues/57#issuecomment-255208477 (already ask if is possible to use conditional routing for SSL pass through)

**Comment by Sushant** [comment]
@aledbf 
Right. For a moment, I thought I could get away with having two instances of the Nginx ingress controller on the same cluster (1 with and 1 w/o SSL passthrough). But then I would need something like this: https://github.com/kubernetes/contrib/pull/1880 to support both types of apps.

**Comment by sandys** [comment]
hi manuel
thanks! btw, the nginxinc bug was also filed by me ;)

do you think we can get nginx 1.5 as the ingress controller and see if tcp
pass through is working? if you check the k8s slack channels... there are a
lot of people who want to do this (we dont want to move our ssl termination
to ingress... because that would mean porting over our configs).

if all i can do is use ingress as a dumb pipe that passes ssl (and source
ip preservation) to my nginx pods... ill be a very happy man!

**Comment by bprashanth** [comment]
there's a proposal open for tcp ingress, and a bug open for multiple tls modes. We should combine those two to give us what we want. Please comment on  https://github.com/kubernetes/kubernetes/issues/32928 and https://github.com/kubernetes/kubernetes/pull/25821 to make sure it does what this bug requires, and re-open if we can't leverage those features for passthrough

**Comment by sandys** [comment]
@bprashanth im commenting here - because i dont want to sound stupid and pollute one of the other bugs. in 25821 - the load balancers that everyone is talking about is cloud based load balancers. for example the readme mentions all the rackspace, vsphere, etc.... but i dont see haproxy/nginx anywhere.

is that going to be added to the bug ? because clearly this bug (and all the people in here) are talking about ingress with a K8s loadbalancer - nginx or haproxy or one of the alternatives

**Comment by bprashanth** [comment]
you can use any controller behind an ingress. If we surface the behavior in the ingress, it doesn't matter what fulfills it. We will still need N bugs for "implement ssl passthrough in [backend of choice] ingress" , but typically that's after "surface ssl passthrough in ingress".

**Comment by sandys** [comment]
@bprashanth sorry to bother you again, but I also am following https://github.com/nginxinc/kubernetes-ingress/issues/57

Do you need to "surface" anything if we are talking nginx or haproxy ingress ? It may true for the different cloud based ingresses/load balancers. I'm trying to figure out whether bug 25821 is blocking the implementation of all these various features (L4, SSL passthrough) for nginx or haproxy ingresses. 

For example, nginx version need to be upgraded in the ingress controller to take advantage of  ngx_stream_ssl_preread_module, but im not sure if that will be done until 25821 is done.. which in turn is blocking for the superset of all the various cloud services.

**Comment by aledbf** [comment]
@sandys this is already working. Please wait while the new version is published in a new repository in the kubernetes organization.

This is how it works: 
1. create an Ingress with the annotation `ingress.kubernetes.io/ssl-passthrough: "true"`

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.kubernetes.io/ssl-passthrough: "true"
  name: kubernetes
  namespace: default
spec:
  rules:
  - host: kubernetes.foo-bar.com
    http:
      paths:
      - backend:
          serviceName: kubernetes
          servicePort: 443
  tls:
  - hosts:
    - kubernetes.foo-bar.com
```

that's it :)

That annotation generates this code in the configuration file:
````
stream {
    # map FQDN that requires SSL passthrough
    map $ssl_preread_server_name $stream_upstream {
        foo-898.bar.com         default-kubernetes-443-0-pt;
        foo-997.bar.com         default-echoheaders-x-80-1-pt;
        kubernetes.foo-bar.com         default-kubernetes-443-2-pt;
        # send SSL traffic to this nginx in a different port
        default                         nginx-ssl-backend;
    }

    log_format log_stream '$remote_addr [$time_local] $protocol [$ssl_preread_server_name] [$stream_upstream] '
        '$status $bytes_sent $bytes_received $session_time';

    access_log /var/log/nginx/access.log log_stream;
    error_log  /var/log/nginx/error.log;

    # configure default backend for SSL
    upstream nginx-ssl-backend {
        server 127.0.0.1:442;
    }

    upstream default-kubernetes-443-0-pt {
        server 172.17.4.99:443;
    }
    upstream default-echoheaders-x-80-1-pt {
        server 10.2.18.4:8080;
    }
    upstream default-kubernetes-443-2-pt {
        server 172.17.4.99:443;
    }

    server {
        listen                  443;
        proxy_pass              $stream_upstream;
        ssl_preread             on;
    }

# TCP services
# UDP services

}
```


> For example, nginx version need to be upgraded

Is a more complicated than that but this is the PR that I will open to [update nginx](https://github.com/kubernetes/contrib/compare/master...aledbf:enable-stream?expand=1)

[Here's](https://github.com/aledbf/ingress-controller/blob/master/backends/nginx/rootfs/etc/nginx/template/nginx.tmpl) the new template

**Comment by sandys** [comment]
@aledbf you are a lifesaver ! 
just to double check because we have been blocked on this 

- does this need a special version of kubernetes ?
- i assume the public DNS of kubernetes.foo-bar.com is pointing to the host with the ingress controller ?
- can i do both port 80 and port 443 ? 
- the way it works for us (atleast in the past with nginx) is that we redirect http -> https. so port 80 gets rewritten to port 443. In this particular case, since the ingress controller will pass it through to the nginx ... will the rewrite still work ?

thanks! this is awesome. Just in time for Diwali ..

**Comment by aledbf** [comment]
> does this need a special version of kubernetes ?

No

> i assume the public DNS of kubernetes.foo-bar.com is pointing to the host with the ingress controller ?

Yes (or just use your /etc/hosts file :) )

> can i do both port 80 and port 443 ?
> the way it works for us (atleast in the past with nginx) is that we redirect http -> https. so port 80 gets rewritten to port 443. In this particular case, since the ingress controller will pass it through to the nginx ... will the rewrite still work ?

Yes but you need to add a redirect annotation ([here's](https://github.com/kubernetes/contrib/issues/1884#issuecomment-254916222) the syntax
Why is this annotation required? I cannot decide for the user what behavior is the "correct" one so if you don't add the annotation the port 80 will not be redirected automatically.
Writing something like:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.kubernetes.io/ssl-passthrough: "true"
  name: kubernetes
  namespace: default
spec:
  backend:
    serviceName: kubernetes
    servicePort: 443
  tls:
  - hosts:
    - kubernetes.foo-bar.com
```

means traffic in port 80 for host `kubernetes.foo-bar.com` redirect to 443

**Comment by sandys** [comment]
@aledbf why do i need to annotate it at all ?

can i not let the traffic go through the port 80 of ingress controller to my nginx pod and then get a 301 redirect ?
I dont necessarily want to do this at the ingress controller level. I actually have many, many redirects in my nginx pod.. and i was wondering if I can simply use my ingress controller to pass traffic to my actual nginx pods.

will that work ? that's the configuration im looking for.

**Comment by aledbf** [comment]
> can i not let the traffic go through the port 80 of ingress controller to my nginx pod and then get a 301 redirect ?

I need to test that. Let me see if it works

**Comment by sandys** [comment]
thanks! really appreciate it!

**Comment by paultiplady** [comment]
@aledbf did this get merged into the k8s ingress repo yet? I didn't see a PR from your repository, and the config seems similar to yours, but not sure if it's the same.

**Comment by aledbf** [comment]
@paultiplady yes. [Here's the code](https://github.com/kubernetes/ingress/blob/master/core/pkg/ingress/annotations/sslpassthrough/main.go)

**Comment by vladkras** [comment]
@aledbf your link is 404

**Comment by AlexProfi** [comment]
Hello Is there an examp;e of all config infrastructure for ingress controller set up as daemon without external lb and deployments with pods that will be terminate ssl.
Ingress contoller must don't terminate ssl pnly pass it to pods on other worker nodes
